README
The Monkeysphere Project
------------------------

The Monkeysphere project's goal is to extend OpenPGP's web of trust to
new areas of the Internet to help us securely identify each other
while we work online.

Specifically, monkeysphere currently offers a framework to leverage
the OpenPGP web of trust for OpenSSH authentication.

In other words, it allows you to use secure shell as you normally do,
but to identify yourself and the servers you administer or connect to
with your OpenPGP keys. OpenPGP keys are tracked via GnuPG, and
monkeysphere manages the known_hosts and authorized_keys files used by
OpenSSH for authentication, checking them for cryptographic validity.

Dependencies
------------

Monkeysphere depends on:

 * GnuPG >= 2.1.11
 * Perl
 * Perl's Crypt::OpenSSL::RSA module
 * lockfile-progs or procmail's lockfile
 * Bash
 * OpenSSH's ssh-keygen utility (ideally >= 6.0)


Changelog
monkeysphere (0.43) unstable; urgency=medium

  * Depend on a modern version of GnuPG (>= 2.1.11) for --export-ssh-key
  * Depend on OpenSSH's ssh-keygen directly for most SSH fingerprints
  * Depend on OpenSSH >= 6.0 for ed25519 and "sshd -T"
  * Use runuser instead of su
  * Support Ed25519 authentication-capable subkeys for users
  * Use https for all outbound links
  * Clean up spelling
  * Use 3072 bits for RSA keys everywhere by default
  * Provide clearer error message for PEM2OPENPGP_NEWKEY
    (Closes: #906755)
  * Avoid locking out users unnecessarily
    (Closes: #897366)

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Wed, 23 Jan 2019 17:42:19 -0500

monkeysphere (0.42) unstable; urgency=medium

  * bugfix release:
  * use --send-keys instead of --send (closes: #908228)
  * tests: Ensure that stale sockets don't fail socat (Closes: #899060)
  * Remove deprecated option from test sshd config (Closes: #902320)
  * write old-style PEM files to unbreak test suite (Closes: #909700)
  * yet more colon fixes that escaped previous inspections
  * fix more gnupg2 colons changes (Closes: #902367)
  * Remove RSAAuthentication from test ssh config (Closes: #902318)
  * clean up test suite failures when built against newer GnuPG
  * use generic compiler (closes: #883015)
  * make print_date_from_seconds_since_the_epoch deal better with bad input

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Tue, 16 Oct 2018 11:38:39 -0400

monkeysphere (0.41) unstable; urgency=medium

  * pem2openpgp now includes issuer fingerprint subpacket
    in hashed self-sig, more compatible with GnuPG 2.1.16
    (Closes: #846554)
  * avoid blocking for entropy during test suite
    (Closes: #841208)
  * augment test suite for id certifier with a subkey, for better realism
  * ensure that attempts to fetch primary key fingerprint only fetch
    primary key fingerprint even if subkey fprs are emitted
    (Closes: #846554)
  * include $CPPFLAGS in agent-transfer build

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Wed, 12 Oct 2016 01:12:27 -0400

monkeysphere (0.40) unstable; urgency=medium

  * bugfix release:
  * get tests to pass against GnuPG 2.1.15
  * build more portably

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Wed, 12 Oct 2016 01:12:27 -0400

monkeysphere (0.39) unstable; urgency=medium

  * avoid warning about unused asprintf return value
  * avoid treating src/share/common as an executable
  * ensure that this works even if SYSSHAREDIR has whitespace
  * Include local build of agent-transfer in $PATH (Closes: #835719)
  * force bash as the shell during su (Closes: #827660)

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Tue, 30 Aug 2016 03:10:53 -0400

monkeysphere (0.38) unstable; urgency=medium

  * make tests work with modern gpg (2.1.x and later) as /usr/bin/gpg
  * this introduces a new binary, agent-transfer, to transfer secret keys
    from gpg-agent to ssh-agent.

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Tue, 17 May 2016 00:44:32 -0400

monkeysphere (0.37) unstable; urgency=medium

  * Bugfix release with minor improvements and dependency accommodations.
  * Test openpgp2ssh functionality (closes MS #6524)
  * use new GnuPG with-colons output
  * accommodate changed behavior of ssh-keygen -F
  * accommodate multiple AuthorizedKeysFile directives
  * deal sanely with empty lines in authorized_user_ids (closes MS #6344)
  * treat non-standard ports properly (closes MS #3402)

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Wed, 06 Aug 2014 13:23:04 -0400

monkeysphere (0.36) unstable; urgency=low

  * keytrans no longer confuses user IDs across different keys (closes MS
    #2682)
  * fetch all available keys from keyserver instead of first 5 (closes MS
    #1046)
  * enable openpgp2pem for keytrans (Closes: #698383)
  * enable openpgp2spki as well

  [ Jonas Smedegaard ]
  * m-a gpg-cmd now takes its arguments as separate parameters, not as a
    single string.

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Thu, 11 Jul 2013 13:19:45 -0400

monkeysphere (0.35) upstream;

  * Remove reference to USE_VALIDATION_AGENT.
  * Fix ssh_proxycommand for marginal hosts (closes MS #2593)
  * GnuPG should always behave as --fixed-list-mode (closes MS #2587)

 -- Jameson Rollins <jrollins@finestructure.net>  Fri, 29 Oct 2010 20:21:54 -0400

monkeysphere (0.34) upstream;

  * fix keys-for-user so that it outputs proper authorized_keys lines
    (close MS #2550)
  * refactor key processing for key files, greatly reducing redundant code
    paths
  * update authorized_keys and known_hosts in temp filess that are
    atomically moved into place
  * don't fail if authorized_keys file not already present (Closes: 600644)
  * document CHECK_KEYSERVER in monkeysphere-authentication man page
    (close MS #2556)

 -- Jameson Rollins <jrollins@finestructure.net>  Tue, 26 Oct 2010 10:27:01 -0400

monkeysphere (0.33) upstream;

  [ Daniel Kahn Gillmor ]
  * defaulting MONKEYSPHERE_HASH_KNOWN_HOSTS to false
    (closes MS #2483)

  [ Jameson Rollins ]
  * fix security vulnerability is parsing userids in
    monkeysphere-authentication keys-for-user (Closes: #600304)
  * fix failure after first invalid key in monkeysphere-authentication
    keys-for-user (closes MS #2545)
  * ignore command options in monkeysphere-authentication keys-for-user

 -- Jameson Rollins <jrollins@finestructure.net>  Fri, 15 Oct 2010 18:05:18 -0400

monkeysphere (0.32) upstream;

  [ Jameson Rollins ]
  * Fix specification of install paths in all scripts and man pages
    (closes MS #2491)
  * Fix need for single argument to gpg_sphere (thanks Clint)
    (closes MS #442)
  * specify LC_ALL=C for all gpg calls
    (closes MS #2496)

  [ Micah Anderson ]
  * fix monkeysphere-host revoke-key, which never worked properly :(
  * add some debug output to monkeysphere-host publish-key
    (closes MS #2289)

  [ Clint Adams ]
  * add support for options to the authorized User IDs file.  Options that

ssh-agent-proto.h
/* see PROTOCOL.agent in openssh repo */

/* 3.2 Requests from client to agent for protocol 2 key operations */

#define SSH2_AGENTC_REQUEST_IDENTITIES 11
#define SSH2_AGENTC_SIGN_REQUEST 13
#define SSH2_AGENTC_ADD_IDENTITY 17
#define SSH2_AGENTC_REMOVE_IDENTITY 18
#define SSH2_AGENTC_REMOVE_ALL_IDENTITIES 19
#define SSH2_AGENTC_ADD_ID_CONSTRAINED 25

/* 3.3 Key-type independent requests from client to agent */

#define SSH_AGENTC_ADD_SMARTCARD_KEY 20
#define SSH_AGENTC_REMOVE_SMARTCARD_KEY 21
#define SSH_AGENTC_LOCK 22
#define SSH_AGENTC_UNLOCK 23
#define SSH_AGENTC_ADD_SMARTCARD_KEY_CONSTRAINED 26

/* 3.4 Generic replies from agent to client */

#define SSH_AGENT_FAILURE 5
#define SSH_AGENT_SUCCESS 6

/* 3.6 Replies from agent to client for protocol 2 key operations */

#define SSH2_AGENT_IDENTITIES_ANSWER 12
#define SSH2_AGENT_SIGN_RESPONSE 14

/* 3.7 Key constraint identifiers */

#define SSH_AGENT_CONSTRAIN_LIFETIME 1
#define SSH_AGENT_CONSTRAIN_CONFIRM 2

/*
All protocol messages are prefixed with their length in bytes, encoded
as a 32 bit unsigned integer. Specifically:

	uint32			message_length
	byte[message_length]	message


RSA keys may be added with this request: 

	byte			SSH2_AGENTC_ADD_IDENTITY or
				SSH2_AGENTC_ADD_ID_CONSTRAINED
	string			"ssh-rsa"
	mpint			rsa_n
	mpint			rsa_e
	mpint			rsa_d
	mpint			rsa_iqmp
	mpint			rsa_p
	mpint			rsa_q
	string			key_comment
	constraint[]		key_constraints


        */

